lea/runner.py

```diff
@@ -131,6 +133,74 @@ def _expand_query(query):
+            if with_clause and isinstance(with_clause, sqlglot.exp.With):
+                for cte in with_clause.expressions:
+                    if isinstance(cte, sqlglot.exp.CTE):
+                        cte_name = cte.alias
+                        cte_query = cte.this.sql(dialect=self.client.sqlglot_dialect)
+                        ctes[cte_name] = cte_query
+
+                # Extract the main query (the part after the CTEs)
+                if "expression" in ast.args:
+                    main_query = ast.args["expression"].sql(dialect=self.client.sqlglot_dialect)
+                else:
+                    # If 'expression' is not in args, use the entire SELECT statement
+                    main_query = ast.sql(dialect=self.client.sqlglot_dialect)
+
+        self.log("Split query result:")
+        self.log(f"Number of CTEs found: {len(ctes)}")
+        for cte_name, cte_query in ctes.items():
+            self.log(f"CTE '{cte_name}': {cte_query[:100]}...")
+
+        return ctes, main_query
+
+    def _materialize_ctes_and_view(self, view):
+        self.log(f"Processing view: {view.key[0]}.{view.key[1]}")
+        self.log(f"Original query:\n{view.query}")
+
+        try:
+            ctes, main_query = self._split_query(view.query)
+        except Exception as e:
+            self.log(f"Error splitting query for {view.key[0]}.{view.key[1]}: {str(e)}")
+            self.log(f"Exception type: {type(e)}")
+            self.log(f"Exception args: {e.args}")
+            raise
+
+        for cte_name, cte_query in ctes.items():
+            try:
+                materialized_table = f"{view.key[0]}.{view.key[1]}__{cte_name}"
+                self.log(f"Attempting to materialize CTE: {materialized_table}")
+                self.log(f"CTE Query: {cte_query}")
+                cte_view = InMemorySQLView(
+                    key=tuple(materialized_table.split(".")), query=cte_query, client=self.client
+                )
+                self.client.materialize_view(cte_view)
+                self.log(f"Successfully materialized CTE: {materialized_table}")
+            except Exception as e:
+                self.log(f"Error materializing CTE {cte_name}: {str(e)}")
+                self.log(f"Exception type: {type(e)}")
+                self.log(f"Exception args: {e.args}")
+                raise
+
+        try:
+            self.log(f"Attempting to materialize main view: {view.key[0]}.{view.key[1]}")
+            self.log(f"Main Query: {main_query}")
+            updated_view = InMemorySQLView(key=view.key, query=main_query, client=self.client)
+            self.client.materialize_view(updated_view)
+            self.log(f"Successfully materialized main view: {view.key[0]}.{view.key[1]}")
+        except Exception as e:
+            self.log(f"Error materializing main view {view.key[0]}.{view.key[1]}: {str(e)}")
+            self.log(f"Exception type: {type(e)}")
+            self.log(f"Exception args: {e.args}")
+            raise
```

Comment:
I'm curious about what happens in terms of execution here. Ideally, we don't want each CTE to run from scratch. We want each CTE to run after its dependencies are materializing. This way, materializing a query with its CTEs doesn't cost more than it does without the splitting process (there will be more stuff to write, but the compute should be the same). Is this something you took into account?

Reply:
Thank you for highlighting this important point, Max. You're right, the current implementation doesn't optimize CTE execution for dependencies. I'll work on implementing a dependency-aware materialization process to ensure efficient execution. This will involve creating a dependency graph for CTEs, using sorting trick to determine the correct execution order, and materializing CTEs only once. I'll update the PR with these changes soon.